tests/test_numpy.py

```diff
@@ -0,0 +1,84 @@
+import numpy as np
+from parameterized import parameterized
+
+from tests import unittest
+
+import jmespath
+import jmespath.functions
+
+
+class TestNumpyNumeric(unittest.TestCase):
+    def setUp(self):
+        self.data = {
+            "a": {
+                "data": np.array([[1,2],[3,4]])
+            },
+            "b": {
+                "data": np.array([[1,2],[3,4]])
+            },
+            "c": {
+                "data": np.array([[2,2],[3,4]])
+            }
+        }
+
+    @parameterized.expand([
+        ["self", "@", lambda data: data],
+        ["get", "a.data", lambda data: data["a"]["data"]],
+        ["slice_horizontal", "a.data[1][:]", lambda data: data["a"]["data"][1,:]],
+        ["slice_vertical", "a.data[:][1]", lambda data: data["a"]["data"][:,1]],
+        ["compare_self", "a.data == a.data", lambda _: True],
+        ["compare_same", "a.data == b.data", lambda _: True],
+        ["compare_other", "a.data == c.data", lambda _: False],
+        ["compare_literal_scalar", "a.data[0][0] == `1`", lambda _: True],
+        ["compare_literal_slice", "a.data[1][:] == `[3, 4]`", lambda _: True],
+        ["compare_literal", "a.data == `[[1,2],[3,4]]`", lambda _: True],
+    ])
+    def test_search(self, _, query, expected):
+        result = jmespath.search(query, self.data)
+        np.testing.assert_equal(result, expected(self.data))
+
+
+class TestNumpyStr(unittest.TestCase):
+    def setUp(self):
+        self.data = {
+            "a": { 
+                "data": np.array(
+                    [
+                        ["test", "messages"],
+                        ["in", "numpy"]
+                    ]
+                )
+            },
+            "b": { 
+                "data": np.array(
+                    [
+                        ["test", "messages"],
+                        ["in", "numpy"]
+                    ]
+                )
+            },
+            "c": { 
+                "data": np.array(
+                    [
+                        ["test", "messages"],
+                        ["other", "numpy"]
+                    ]
+                )
+            }
+        }
+
+    @parameterized.expand([
+        ["self", "@", lambda data: data],
+        ["get", "a.data", lambda data: data["a"]["data"]],
+        ["slice_horizontal", "a.data[1][:]", lambda data: data["a"]["data"][1,:]],
+        ["slice_vertical", "a.data[:][1]", lambda data: data["a"]["data"][:,1]],
+        ["compare_self", "a.data == a.data", lambda _: True],
+        ["compare_same", "a.data == b.data", lambda _: True],
+        ["compare_other", "a.data == c.data", lambda _: False],
+        ["compare_literal_scalar", "a.data[0][0] == 'test'", lambda _: True],
+        ["compare_literal_slice", "a.data[1][:] == ['in', 'numpy']", lambda _: True],
+        ["compare_literal", "a.data == [['test', 'messages'],['in', 'numpy']]", lambda _: True],
+    ])
+    def test_search(self, _, query, expected):
```

Comment:
maybe still receive the first parameter as `query_name` or similar and use it in the error message of `assert_equal` (assuming it allows you to give a custom message?)

Reply:
Typically the error message gives details of where in the data the assertion failed instead of the test name. It'll raise something like this on error:

![image](https://github.com/calgray/jmespath.py/assets/7478405/f5b93bc0-87ac-483f-ace3-5ce67d05fb7f)

Reply:
OK, there seems to be no information loss then. The data involved in the failure is still shown by default, at least from what I can see in your example above and local tests (I removed the traceback for readability):

```python
>>> np.testing.assert_equal(True, False)
AssertionError: 
Items are not equal:
 ACTUAL: True
 DESIRED: False
>>>
>>> np.testing.assert_equal(True, False, err_msg="blah")
AssertionError: 
Items are not equal: blah
 ACTUAL: True
 DESIRED: False
```

```python
>>> np.testing.assert_equal(a1, [1, 2, 4])
AssertionError: 
Arrays are not equal

Mismatched elements: 1 / 3 (33.3%)
Max absolute difference: 1
Max relative difference: 0.25
 x: array([1, 2, 3], dtype=int32)
 y: array([1, 2, 4])
>>>
>>> np.testing.assert_equal(a1, [1, 2, 4], err_msg="blah")
AssertionError: 
Arrays are not equal
blah
Mismatched elements: 1 / 3 (33.3%)
Max absolute difference: 1
Max relative difference: 0.25
 x: array([1, 2, 3], dtype=int32)
 y: array([1, 2, 4])
```

I'd then say let's add it, it would make it a tad easier to catch potential errors later on.